### PR TITLE
Add warning on underscored variable access

### PR DESCRIPTION
**Pull request by willpenington**
Adds a warning when a variable prefixed with an underscore is accessed.
Message and code are based on the warning for repetition of the same
underscored variable in a match clause. Implements issue #3386

### CODE REVIEW

**Comment by willpenington** [comment]
Building Elixir with this option enabled in Travis CI picks up __impl__ from the protocol implementation. Is this something that needs to be special cased?

**Comment by josevalim** [comment]
Do you think it would be possible to move this check to `elixir_scope:translate_var`?

We would need to fix `__impl__` if we plan to move ahead with this. :) Thank you!

**Comment by willpenington** [comment]
I've moved the code across into `elixir_scope.erl` and ignored `__init__` and the automatically generated variables.

File: lib/elixir/src/elixir_scope.erl [hunk 1]
```diff
@@ -75,6 +76,20 @@ warn_underscored_var(Line, File, Name, Kind) ->
       ok
   end.
 
+warn_underscored_var_access(Line, File, Name) ->
+  case atom_to_list(Name) of
+    "__impl__" ->
+      ok; %% Variable needed for protocols
+    "_rewrite" ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_scope.erl]: We should fix those. We probably don't need to name it `_rewrite` nor `__impl__`. :)

**Comment by willpenington** [comment]
Done.

**Comment by josevalim** [comment]
Perfect, thank you!

File: lib/elixir/src/elixir_scope.erl [hunk 2]
```diff
@@ -161,4 +172,10 @@ format_error({unused_match, Name, Kind}) ->
...
                 "to the same value. If this is the intended behaviour, please "
                 "remove the leading underscore from the variable name, otherwise "
-                "give the variables different names", [Name, context_info(Kind), Name]).
+                "give the variables different names", [Name, context_info(Kind), Name]);
+
+format_error({underscore_var_access, Name}) ->
+  io_lib:format("the underscored variable \"~ts\" is used in the body of the "
+                "function. A leading underscore indicates that the value of "
+                "the variable should be ignored. If this is intended please "
+                "rename the variable remove the underscore", [Name]).
```
Comment (henrik) [inline lib/elixir/src/elixir_scope.erl]: Typo: "rename the variable **to** remove the underscore"

**Comment by josevalim** [comment]
@henrik i will fix it as a i merge it locally. Thanks @willpenington !

**Pull request closed without merging**